pkg/collector/corechecks/snmp/internal/report/report_bandwidth_usage.go

```diff
@@ -97,18 +98,76 @@ func (ms *MetricSender) sendBandwidthUsageMetric(symbol profiledefinition.Symbol
 	}
 	usageValue := ((octetsFloatValue * 8) / (float64(ifSpeed))) * 100.0
 
-	sample := MetricSample{
-		value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeCounter, Value: usageValue},
-		tags:       tags,
-		symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
-		forcedType: profiledefinition.ProfileMetricTypeCounter,
-		options:    profiledefinition.MetricsConfigOption{},
+	interfaceID := ms.deviceID + ":" + fullIndex + "." + usageName
+	err = ms.calculateRate(interfaceID, ifSpeed, usageValue, usageName, tags)
+	if err != nil {
+		return err
 	}
+	return nil
+}
+
+/*
+calculateRate is responsible for checking the state for previously seen metric sample to generate the rate from.
+If the ifSpeed has changed for the interface, the rate will not be submitted (drop the previous sample)
+*/
+func (ms *MetricSender) calculateRate(interfaceID string, ifSpeed uint64, usageValue float64, usageName string, tags []string) error {
+	ms.interfaceRateMap.mu.RLock()
+	interfaceRate, ok := ms.interfaceRateMap.rates[interfaceID]
+	ms.interfaceRateMap.mu.RUnlock()
+
+	// current data point has the same interface speed as last data point
+	if ok && interfaceRate.ifSpeed == ifSpeed {
+		// calculate the delta, taken from pkg/metrics/rate.go
+		// https://github.com/DataDog/datadog-agent/blob/ecedf4648f41193988b4727fc6f893a0dfc3991e/pkg/metrics/rate.go#L37
+		currentTimestamp := TimeNow()
+		delta := (usageValue - interfaceRate.previousSample) / (currentTimestamp - interfaceRate.previousTs)
+		// update the map previous as the current for next rate
+		interfaceRate.previousSample = usageValue
+		interfaceRate.previousTs = currentTimestamp
+		ms.interfaceRateMap.mu.Lock()
+		ms.interfaceRateMap.rates[interfaceID] = interfaceRate
```

Comment:
Should we have a mechanism that cleans old/unused interfaceID entries, or are we confident that `interfaceRateMap` will never be "too big" ?

Reply:
Hm, this is a good point - I'm not sure of two things that could impact this:
* How often will the agent be restarted (I'm assuming this will reinitialize the check and thus also the map)
* How common it is for customers to remove devices and/or interfaces and (and thus how large the map can get for the average/max number of devices and interfaces they will monitor)

Is there a dashboard somewhere or some analytics to see across customers how many devices we support for each customer/something to see their agent restarts all in one place? I know of the dashboard you shared for the NDM environment, but something aggregating across multiple customers may not be possible?

Reply:
I'm not sure we have such a dashboard unfortunately 🙁 
I guess for the map to grow too big the customer would have to add and remove thousands of devices, so this might not be an issue!

Reply:
@zoedt 

I think this is a good point, I think it might be safer to protect against the map growing due to interfaces that are not used anymore. If this situation happen, this is a memory leak, which can cause Agent to crash at some point.

I see multiple reasons why the map can grow:
- if an Agent is not restarted for long time
- user adds and remove a lot of interfaces frequently
- there are possibly devices that can have virtual interfaces dynamically created with indexes that are random

A possible way to protect against this is to have a clean up function called in `MetricSender.ReportMetrics()` here https://github.com/DataDog/datadog-agent/blob/7abd458e82257bc801867d8289e1175977bcebd5/pkg/collector/corechecks/snmp/internal/report/report_metrics.go#L78

Something like this:
```
timeBeforeReportingMetrics := time.Now()

[...] rest of code from MetricSender.ReportMetrics()

ms.interfaceBandwidthState.ClearUnusedBandwidthUsageStates(timeBeforeReportingMetrics)
```

Then, inside `ClearUnusedBandwidthUsageStates`, you can remove all entries that have "previousTsNano" older than `timeBeforeReportingMetrics`.

Reply:
I suppose `timeBeforeReportingMetrics` might erase entries from the previous check run since report metrics happens on each run?

Maybe I can take the time before reporting metrics but subtract one hour?